### PR TITLE
PT-938 | New occurrences form

### PR DESCRIPTION
**Pull request by Sartonon**
## Description :sparkles:

- Updated event wizard
- New language selector so all language versions can be given easily
- Updated second step of event wizard

### Closes :no_good_woman:
**[PT-973](https://helsinkisolutionoffice.atlassian.net/browse/PT-973):** 
**[PT-935](https://helsinkisolutionoffice.atlassian.net/browse/PT-935):** 
**[PT-937](https://helsinkisolutionoffice.atlassian.net/browse/PT-937):** 
**[PT-938](https://helsinkisolutionoffice.atlassian.net/browse/PT-938):** 

### Automated tests :gear:️

- Automated tests still needs to be fixed / updated for the redesign

### Manual testing :construction_worker_man:

- Test heavily by adding many languages and going between wizard step updating and saving data
- Make sure that all the changes are preserved and publishing is working

## Screenshots :camera_flash:

Screenshot 2021-04-28 at 10 00 23
Screenshot 2021-04-28 at 9 59 58
Screenshot 2021-04-28 at 9 59 52

## Additional notes :spiral_notepad:

### CODE REVIEW

**Review by nikomakela — COMMENTED** [review]
I did a very light review - there was so much code to go through. I think I will concentrate more on manual testing. The code looked good, didn't find any typos from translations, not sure about Swedish, etc. Didn't even try to improve the code effectiveness.

At this point when the budget is tight and this feature is hardly needed, I would say that we should focus on testing (manual functional testing and unit tests as automated).

**Comment by nikomakela** [comment]
These fields shouldn't be rendered before we have that feature in API. 
image

**Comment by nikomakela** [comment]
Occurrence before enrolment starts:

`[GraphQL error]: Message: Start time cannot be before the enrolment ends, Location: [object Object], Path: addOccurrence`

image

**Comment by nikomakela** [comment]
Just a note: the UI comes a bit laggy when there are 2 languages. However, I don't see much difference when there are 3 languages - it's as laggy as with 2 languages.

image

**Comment by nikomakela** [comment]
When moving from step1 to step2: 

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in EditEventPage (created by Context.Consumer)

**Comment by nikomakela** [comment]
Occurrence event location can be changed when it's an internet event

image


image

**Comment by nikomakela** [comment]
> These fields shouldn't be rendered before we have that feature in API.
> image

Same thing with "Enrolment to event is required":
image

Of course we can try to bite a huge piece  and implement them :)

File: package.json [hunk 1]
```diff
@@ -57,6 +57,7 @@
     "react-scripts": "3.4.0",
     "react-table": "^7.2.2",
     "react-toastify": "^5.5.0",
+    "react-use": "^17.2.3",
```
[the remark this answers is not in the document]
Comment (Sartonon) [reply package.json:60]: Only used it for a while for one hook but it isn't needed so this can be deleted

**Comment by Sartonon** [comment]
> Occurrence before enrolment starts:
> 
> `[GraphQL error]: Message: Start time cannot be before the enrolment ends, Location: [object Object], Path: addOccurrence`
> 
> image
> 
> And there is that enrolment end days -field also that forces the ending of enrolment to be before the occurrence also.
> 
> So... this works as it should
> 
> image
> 
> A validation to UI needs to be added. Also the error message given by API could be shown somehow, but that's another task.

Yeah known issue.. I will add UI validation

**Comment by Sartonon** [comment]
> When moving from step1 to step2:
> 
> > Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
> > in EditEventPage (created by Context.Consumer)

This needs more investigation, these warnings have been issue from the beginning and can be time consuming to resolve. Fortunately they shouldn't cause issues but are annoying in the logs

**Comment by Sartonon** [comment]
> Occurrence event location can be changed when it's an internet event
> 
> image
> 
> image

Yeah this needs to be handled somehow

**Comment by Sartonon** [comment]
> > These fields shouldn't be rendered before we have that feature in API.
> > image
> 
> Same thing with "Enrolment to event is required":
> image
> 
> Of course we can try to bite a huge piece and implement them :)

Yup, needs to be removed

**Comment by nikomakela** [comment]
Publish single event fails

image

> [GraphQL error]: Message: Expecting value: line 1 column 1 (char 0), Location: [object Object], Path: publishEventMutation

> {"errors":[{"message":"Expecting value: line 1 column 1 (char 0)","locations":[{"line":2,"column":3}],"path":["publishEventMutation"],"extensions":{"code":"GENERAL_ERROR"}}],"data":{"publishEventMutation":null}}

```
curl 'http://localhost:8081/graphql' \
  -H 'Connection: keep-alive' \
  -H 'Pragma: no-cache' \
  -H 'Cache-Control: no-cache' \
  -H 'sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="90", "Google Chrome";v="90"' \
  -H 'accept: */*' \
  -H 'authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6ImU0Mzg4MzBiYzBkMTQ5MzMyMDFjYzhmOWFjNmNmZDY3In0.eyJpc3MiOiJodHRwOi8vdHVubmlzdGFtby1iYWNrZW5kOjg4ODgvb3BlbmlkIiwic3ViIjoiYTFjNjNkYTYtNWVlZi0xMWViLTllMmQtMDI0MmFjMTMwMDAzIiwiYXVkIjoiaHR0cHM6Ly9hcGkuaGVsLmZpL2F1dGgvcGFsdmVsdXRhcmpvdGluIiwiZXhwIjoxNjE5NzA0OTQ1LCJpYXQiOjE2MTk3MDIxODUsImF1dGhfdGltZSI6MTYxOTQyNzAyOSwibmFtZSI6Ik5pa28gTGluZHJvb3MiLCJnaXZlbl9uYW1lIjoiTmlrbyIsImZhbWlseV9uYW1lIjoiTGluZHJvb3MiLCJuaWNrbmFtZSI6Ik5pa28iLCJlbWFpbCI6Im5pa28uai5tYWtlbGFAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJhbXIiOiJnaXRodWIiLCJhenAiOiJodHRwczovL2FwaS5oZWwuZmkvYXV0aC9wYWx2ZWx1dGFyam90aW4tYWRtaW4iLCJsb2EiOiJsb3ciLCJodHRwczovL2FwaS5oZWwuZmkvYXV0aCI6WyJwYWx2ZWx1dGFyam90aW4iXX0.WtkVRAXIP8muEhPiYJRThLe5Vs1bSrur98w6rZvVzDdDbFaV_as821ojRenFscnRQaB3xaz-pJi96sddknF2IPfVEi1XPTQXWle3iRsEr3y-aBlS0ZcO-2_Pxm3-Zc0iJgStT1Coks1NMVt00j6xgzJpbJqY0FNqH9nvbWjvu7fcUEJDJ8hvmQUXWVlAR7qaS2bx2QSztZ5Ak-AX2R7ERF5UyTEeC4vhlpftEnIf4214tAK9-hh7sm4vC73rfS6vd5wqAW6t_5wCYoAFEYiT1TzUxKDjl0wYgZ4_oUa12RUthNZGeQLtT6j1v8B1cAvYcY7NvKjXgxsJ7fCdRCYQurZ0YeD18Tg-Ap5loNF__0hTGTHF6CbIdcxmp3YIP7-04-35LY1bUs5cRrm3LldnMzOS5513rTeuCo8UylFg52G9tilZ3WEq94VHdqD0XzpRfPGRWTYzIhan_RjhIGQNe8y9Izb-zW_QpiP0VExlgTKwYa4A28Gzhk0T2CO7o_ha216VnQqbWxmpm_AW0crsSNFEE4hFFB28qR7s1e2bVyAOJUpay12MjQeYlJvPeMXM8mYZNyevcdo7pRhoRozZhoUNTT1h7cuyuW9Lq9Jq7h1wQJEHgSqwvY-ik2BMSmRhwSvmuawzF7h_OFIvFgsi7GxmvU-DBY9VONJizCsgO4I' \
  -H 'Accept-language: fi' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36' \
  -H 'content-type: application/json' \
  -H 'Origin: http://localhost:3000' \
  -H 'Sec-Fetch-Site: same-site' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Referer: http://localhost:3000/' \
  --data-raw $'{"operationName":"publishSingleEvent","variables":{"event":{"id":"local-kultus:af4r3rtf24","location":{"internalId":"http://localhost:8000/v1/place/tprek:50315/"},"keywords":[{"internalId":"http://localhost:8000/v1/keyword/yso:p10452/"},{"internalId":"http://localhost:8000/v1/keyword/yso:p1481/"}],"offers":[{"isFree":true,"description":{"en":"","fi":"","sv":""},"price":{"en":"","fi":"","sv":""},"infoUrl":null}],"name":{"en":"","fi":"","sv":"Ruotsi"},"description":{"en":"","fi":"","sv":"ruotsi"},"pEvent":{"neededOccurrences":1,"mandatoryAdditionalInformation":false},"shortDescription":{"en":"","fi":"","sv":"Ruotsi"},"organisationId":"T3JnYW5pc2F0aW9uTm9kZTox"}},"query":"mutation publishSingleEvent($event: PublishEventMutationInput\u0021) {\\n  publishEventMutation(event: $event) {\\n    response {\\n      statusCode\\n      body {\\n        id\\n        internalId\\n        publicationStatus\\n        __typename\\n      }\\n      resultText\\n      __typename\\n    }\\n    __typename\\n  }\\n}\\n"}' \
  --compressed
```

**Comment by nikomakela** [comment]
After 2 languages are filled, if I want to remove 1 of them, I deselect the language checkbox and the UI reacts to it. But the data is still there and after saving the form, it will also save the deactivated language.

image

**Comment by Sartonon** [comment]
> After 2 languages are filled, if I want to remove 1 of them, I deselect the language checkbox and the UI reacts to it. But the data is still there and after saving the form, it will also save the deactivated language.
> 
> image
> 
> EDIT: I need to empty the fields manually first and then deselect the language. Then the data is really removed.

Fixed

**Comment by Sartonon** [comment]
> Publish single event fails
> 
> image
> 
> > [GraphQL error]: Message: Expecting value: line 1 column 1 (char 0), Location: [object Object], Path: publishEventMutation
> 
> > {"errors":[{"message":"Expecting value: line 1 column 1 (char 0)","locations":[{"line":2,"column":3}],"path":["publishEventMutation"],"extensions":{"code":"GENERAL_ERROR"}}],"data":{"publishEventMutation":null}}
> 
> ```
> curl 'http://localhost:8081/graphql' \
>   -H 'Connection: keep-alive' \
>   -H 'Pragma: no-cache' \
>   -H 'Cache-Control: no-cache' \
>   -H 'sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="90", "Google Chrome";v="90"' \
>   -H 'accept: */*' \
>   -H 'authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6ImU0Mzg4MzBiYzBkMTQ5MzMyMDFjYzhmOWFjNmNmZDY3In0.eyJpc3MiOiJodHRwOi8vdHVubmlzdGFtby1iYWNrZW5kOjg4ODgvb3BlbmlkIiwic3ViIjoiYTFjNjNkYTYtNWVlZi0xMWViLTllMmQtMDI0MmFjMTMwMDAzIiwiYXVkIjoiaHR0cHM6Ly9hcGkuaGVsLmZpL2F1dGgvcGFsdmVsdXRhcmpvdGluIiwiZXhwIjoxNjE5NzA0OTQ1LCJpYXQiOjE2MTk3MDIxODUsImF1dGhfdGltZSI6MTYxOTQyNzAyOSwibmFtZSI6Ik5pa28gTGluZHJvb3MiLCJnaXZlbl9uYW1lIjoiTmlrbyIsImZhbWlseV9uYW1lIjoiTGluZHJvb3MiLCJuaWNrbmFtZSI6Ik5pa28iLCJlbWFpbCI6Im5pa28uai5tYWtlbGFAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJhbXIiOiJnaXRodWIiLCJhenAiOiJodHRwczovL2FwaS5oZWwuZmkvYXV0aC9wYWx2ZWx1dGFyam90aW4tYWRtaW4iLCJsb2EiOiJsb3ciLCJodHRwczovL2FwaS5oZWwuZmkvYXV0aCI6WyJwYWx2ZWx1dGFyam90aW4iXX0.WtkVRAXIP8muEhPiYJRThLe5Vs1bSrur98w6rZvVzDdDbFaV_as821ojRenFscnRQaB3xaz-pJi96sddknF2IPfVEi1XPTQXWle3iRsEr3y-aBlS0ZcO-2_Pxm3-Zc0iJgStT1Coks1NMVt00j6xgzJpbJqY0FNqH9nvbWjvu7fcUEJDJ8hvmQUXWVlAR7qaS2bx2QSztZ5Ak-AX2R7ERF5UyTEeC4vhlpftEnIf4214tAK9-hh7sm4vC73rfS6vd5wqAW6t_5wCYoAFEYiT1TzUxKDjl0wYgZ4_oUa12RUthNZGeQLtT6j1v8B1cAvYcY7NvKjXgxsJ7fCdRCYQurZ0YeD18Tg-Ap5loNF__0hTGTHF6CbIdcxmp3YIP7-04-35LY1bUs5cRrm3LldnMzOS5513rTeuCo8UylFg52G9tilZ3WEq94VHdqD0XzpRfPGRWTYzIhan_RjhIGQNe8y9Izb-zW_QpiP0VExlgTKwYa4A28Gzhk0T2CO7o_ha216VnQqbWxmpm_AW0crsSNFEE4hFFB28qR7s1e2bVyAOJUpay12MjQeYlJvPeMXM8mYZNyevcdo7pRhoRozZhoUNTT1h7cuyuW9Lq9Jq7h1wQJEHgSqwvY-ik2BMSmRhwSvmuawzF7h_OFIvFgsi7GxmvU-DBY9VONJizCsgO4I' \
>   -H 'Accept-language: fi' \
>   -H 'sec-ch-ua-mobile: ?0' \
>   -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36' \
>   -H 'content-type: application/json' \
>   -H 'Origin: http://localhost:3000' \
>   -H 'Sec-Fetch-Site: same-site' \
>   -H 'Sec-Fetch-Mode: cors' \
>   -H 'Sec-Fetch-Dest: empty' \
>   -H 'Referer: http://localhost:3000/' \
>   --data-raw $'{"operationName":"publishSingleEvent","variables":{"event":{"id":"local-kultus:af4r3rtf24","location":{"internalId":"http://localhost:8000/v1/place/tprek:50315/"},"keywords":[{"internalId":"http://localhost:8000/v1/keyword/yso:p10452/"},{"internalId":"http://localhost:8000/v1/keyword/yso:p1481/"}],"offers":[{"isFree":true,"description":{"en":"","fi":"","sv":""},"price":{"en":"","fi":"","sv":""},"infoUrl":null}],"name":{"en":"","fi":"","sv":"Ruotsi"},"description":{"en":"","fi":"","sv":"ruotsi"},"pEvent":{"neededOccurrences":1,"mandatoryAdditionalInformation":false},"shortDescription":{"en":"","fi":"","sv":"Ruotsi"},"organisationId":"T3JnYW5pc2F0aW9uTm9kZTox"}},"query":"mutation publishSingleEvent($event: PublishEventMutationInput\u0021) {\\n  publishEventMutation(event: $event) {\\n    response {\\n      statusCode\\n      body {\\n        id\\n        internalId\\n        publicationStatus\\n        __typename\\n      }\\n      resultText\\n      __typename\\n    }\\n    __typename\\n  }\\n}\\n"}' \
>   --compressed
> ```

Don't know what is causing this, working for me

**Pull request merged**